[[ Engine ]] Add 'layerClipRect' property to controls

livecodeian opened the pull request:
This patch adds a new property 'layerClipRect' to all controls.

The property defines a clipping rectangle in card co-ordinates which
is applied both when the control is rendered normally, and also to
the accelerated rendering layer-clip when accelerated rendering is
turned on.

The property means that the restriction of only top-level objects
being able to have dynamic layers in accelerated rendering mode
is not quite such a problem as it is now possible to have a group
clipping like affect applied to top-level objects.

*NOTE* This PR merges branch https://github.com/runrevmark/livecode/tree/accelrender-layer_clip_rect into PR https://github.com/livecode/livecode/pull/7006

montegoulding commented:
@livecodeian I think this needs to be rebased on #7006 or something... or perhaps just one PR?

runrevmark commented:
@livecodeian I think it would be neater if you just picked this commit into the container layermode PR :)

runrevmark commented:
@livecodeian Could you try and neaten up the commits here a bit  - i.e. fold the release note/dictionary entry into the main layerClipRect commit and then see if you can make it depend on #7006 (in particular get rid of the merge commit as that will confuse the history a bit!).

livecodeian committed:
[[ Engine ]] Add 'layerClipRect' property to controls

This patch adds a new property 'layerClipRect' to all controls.

The property defines a clipping rectangle in card co-ordinates which
is applied both when the control is rendered normally, and also to
the accelerated rendering layer-clip when accelerated rendering is
turned on.

livecodeian commented:
ok, I've rebased this on #7006 now and folded the docs into the main commit

runrevmark commented on docs/dictionary/property/layerClipRect.lcdoc:
```diff
@@ -0,0 +1,38 @@
+A rectangle in card coordinates.
+
+Description:
+Use the <layerClipRect> property to define a clipping rectangle in card
```
I suggest:

Use the <layerClipRect> property to clip an object's display to a rectangle. The clipping
rectangle only changes what part of the object is rendered, it has no effect on interaction;
in particular, mouse events will still occur as they would without it being set.

runrevmark commented on docs/notes/feature-layercliprect.md:
```diff
@@ -0,0 +1,13 @@
+
+A new property 'layerClipRect' has been added to all controls.
+
+The property defines a clipping rectangle in card co-ordinates which
```
I suggest just copying the revised description above and ditching the second paragraph here - as container layerMode lifts the restriction in a more general way.

runrevmark commented:
@livecode-vulcan review ok 6300ef6

livecode-vulcan commented:
💙 review by @runrevmark ok 6300ef6c87a8180028e23c3821e2a19f39ee8272

livecode-vulcan commented:
😎 test success 6300ef6c87a8180028e23c3821e2a19f39ee8272

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/511)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/503)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/503)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/503)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/503)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2025)
* try-community-universal-ios-iphoneos12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos12.1/builds/253)
* try-community-universal-ios-iphonesimulator12.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator12.1/builds/252)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2057)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2047)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2045)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2080)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2017)
